Improve `--help` content in both firecracker and jailer

shioyama18 opened the pull request:
Signed-off-by: Shion Yamashita <shioyama1118@gmail.com>

## Reason for This PR

Fix: #1688

## Description of Changes

Improve the following in `--help` content
- Display default values along with the description
- Segregate required and optional arguments
- Sort and align arguments

```
Jailer v0.21.0

required arguments:
  --exec-file <exec-file>   File path to exec into.
  --gid <gid>               The group identifier the jailer switches to after exec.
  --id <id>                 Jail ID.
  --node <node>             NUMA node to assign this microVM to.
  --uid <uid>               The user identifier the jailer switches to after exec.

optional arguments:
  --chroot-base-dir <chroot-base-dir>   The base folder where chroot jails are located. [default: "/srv/jailer"]
  --daemonize                           Daemonize the jailer before exec, by invoking setsid(), and redirecting the standard I/O file descriptors to /dev/null.
  --extra-args <extra-args>             Arguments that will be passed verbatim to the exec file.
  --help                                Show the help message.
  --netns <netns>                       Path to the network namespace this microVM should join.
```

```
Firecracker v0.21.0

optional arguments:
  --api-sock <api-sock>                     Path to unix domain socket used by the API. [default: "/run/firecracker.socket"]
  --config-file <config-file>               Path to a file that contains the microVM configuration in JSON format.
  --help                                    Show the help message.
  --id <id>                                 MicroVM unique identifier. [default: "anonymous-instance"]
  --level <level>                           Set the logger level. [default: "Warning"]
  --log-path <log-path>                     Path to a fifo or a file used for configuring the logger on startup.
  --no-api                                  Optional parameter which allows starting and using a microVM without an active API socket.
  --seccomp-level <seccomp-level>           Level of seccomp filtering (0: no filter | 1: filter by syscall number | 2: filter by syscall number and argument values) that will be passed to executed path as argument. [default: "2"]
  --show-level                              Whether or not to output the level in the logs.
  --show-log-origin                         Whether or not to include the file path and line number of the log's origin.
  --start-time-cpu-us <start-time-cpu-us>   Process start CPU time (wall clock, microseconds).
  --start-time-us <start-time-us>           Process start time (wall clock, microseconds).
```

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

aghecenco commented on src/utils/src/arg_parser.rs:
```diff
@@ -78,17 +78,61 @@ impl<'a> ArgParser<'a> {
+            .into_iter()
+            .map(|argument| argument.1)
+            .filter(|arg| {
+                if is_required {
+                    arg.required
+                } else {
+                    !arg.required
+                }
```
```suggestion
                is_required == arg.required
```

aghecenco commented on src/firecracker/src/main.rs:
```diff
@@ -106,11 +106,13 @@ fn main() {
             Argument::new("start-time-us")
-                .takes_value(true),
+                .takes_value(true)
+                .help("Wall clock time calculated by the jailer that it spent doing its work.")
```
```suggestion
                .help("Process start time (wall clock, microseconds).")
```

aghecenco commented on src/firecracker/src/main.rs:
```diff
@@ -106,11 +106,13 @@ fn main() {
             Argument::new("start-time-cpu-us")
-                .takes_value(true),
+                .takes_value(true)
+                .help("CPU time calculated by the jailer that it spent doing its work."),
```
```suggestion
                .help("Process start CPU time (wall clock, microseconds)."),
```

shioyama18 commented:
@aghecenco 
Thank you for the review. I updated the help message and filter predicates.

shioyama18 commented:
@lauralt 
Thank you for the review. I squashed the commits and fixed all issues mentioned in comments.
For the test, I defined `width` that is longer than the width of the argument and `short_width` that is shorter than the width of the argument, and tested each case with parameters, help message, and default value.

lauralt commented on src/utils/src/arg_parser.rs:
```diff
@@ -150,20 +186,40 @@ impl<'a> Argument<'a> {
-        if self.takes_value {
-            help_builder.push(format!(" <{}>", self.name));
-        }
+        // Add three whitespaces between command and help message for readablity.
```
```suggestion
        // Add three whitespaces between the argument and its help message for readability.
```

shioyama18 replied:
Done

lauralt commented on src/utils/src/arg_parser.rs:
```diff
@@ -150,20 +186,40 @@ impl<'a> Argument<'a> {
     }
 
-    fn format_help(&self) -> String {
+    fn format_help(&self, max_arg_width: usize) -> String {
```
Hmm, for this function context, I don't think `max_arg_width` is an accurate name, it's more of a fixed width than max. I think I prefer the old version (`arg_width` if I remember correctly).

shioyama18 replied:
Done. Thank you for the comment.

shioyama18 commented:
@lauralt 
I updated the commit message and coverage target value, but I am now failing the tests in `test_rate_limiter.py`. Is `MAX_BYTES_DIFF_PERCENTAGE` also a value to be updated?

lauralt commented:
> @lauralt
> I updated the commit message and coverage target value, but I am now failing the tests in `test_rate_limiter.py`. Is `MAX_BYTES_DIFF_PERCENTAGE` also a value to be updated?

No.. that test is failing intermittently, we have an open issue for it: #1809 and a PR that is trying to fix it: #1898. So, another `push -f` should be enough.

Anyway, it looks like you have a conflict in the coverage file. Meantime, there were merged some PRs which increased the coverage target pretty much, so the value added by you is not accurate anymore.
I suggest you to rebase your branch against our master when there are many commits merged since you opened the PR and maybe run locally the coverage test: `tools/devtool test -- integration_tests/build/test_coverage.py::test_coverage`. I did that earlier for your branch (updated with the changes from master) and the new value seems to be: `84.56`, you can try with this one even though in the CI this value varies slightly (you still need to do the rebase).

shioyama18 commented:
@lauralt 
Thank you for the comment, I rebased to master and all tests passed successfully.

lauralt commented:
Hi, @shioyama18! The coverage test is failing again :(. Can you please update the target value to `84.59`? Sorry for that.

shioyama18 commented:
Hi @lauralt, I added commit to update target value to `84.59`.

lauralt commented:
Thanks, one more little thing :D: would be awesome if you could squash the commits into a single one, we don't usually have a separate commit just for the coverage change, sorry again :(.

shioyama18 commented:
No problem. Thank you for your kind review.
